src/storage.rs

```diff
@@ -1,296 +1,494 @@
 //! Ring buffer implementation, that does immutable reads.
 
-use std::cell::UnsafeCell;
-use std::marker;
-use std::ops::{Index, IndexMut};
-use std::sync::Arc;
-use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering, ATOMIC_USIZE_INIT};
+use std::marker::PhantomData;
+use std::num::Wrapping;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
+use std::ptr;
+use std::sync::mpsc::{self, Receiver, Sender};
 
-/// The reader id is used by readers to tell the storage where the last read ended.
-#[derive(Debug)]
-pub struct ReaderId<T> {
-    reader_id: usize,
-    buffer_id: usize,
-    alive: Arc<AtomicBool>,
-    m: marker::PhantomData<T>,
+use parking_lot::Mutex;
+
+#[derive(Clone, Copy, Debug)]
+struct CircularIndex {
+    index: usize,
+    size: usize,
 }
 
-impl<T> ReaderId<T> {
-    /// Create a new reader id
-    pub fn new(reader_id: usize, buffer_id: usize, alive: Arc<AtomicBool>) -> ReaderId<T> {
-        ReaderId {
-            reader_id,
-            buffer_id,
-            alive,
-            m: marker::PhantomData,
+impl CircularIndex {
+    fn new(index: usize, size: usize) -> Self {
+        CircularIndex { index, size }
+    }
+
+    fn at_end(size: usize) -> Self {
+        CircularIndex {
+            index: size - 1,
+            size,
+        }
+    }
+
+    /// A magic value (!0).
+    /// This value gets set when calling `step` and we reached
+    /// the end.
+    fn magic(size: usize) -> Self {
+        CircularIndex::new(!0, size)
+    }
+
+    fn is_magic(&self) -> bool {
+        self.index == !0
+    }
+
+    fn step(&mut self, inclusive_end: usize) -> Option<usize> {
```

Comment:
I'm not a big fan of using magic like !0 to indicate the end of the steps. If inclusive_end == !0 you'll skip an event.

Reply:
much better than the alternative (double special case)